File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        seen = {0: 1}          # prefix_sum -> count of times seen
        total = 0
        result = 0
        for num in nums:
            total += num
            result += seen.get(total - k, 0)   # query BEFORE insert
            seen[total] = seen.get(total, 0) + 1
        return result

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> mapFreq = new HashMap<>();
        mapFreq.put(0,1);
        int count = 0;
        Integer total = 0;
        for(int i = 0; i < nums.length; i++){
            total += nums[i];
            int prefixSum = total - k;
            mapFreq.putIfAbsent(total, 0);
            if (mapFreq.containsKey(prefixSum)){
                count += mapFreq.get(prefixSum);
            }
            mapFreq.put(total, mapFreq.get(total) + 1);   
        }
        return count;
    }
}